pgmpy/models/BayesianNetwork.py

```diff
@@ -712,39 +712,53 @@ def predict(self, data, algo=None, stochastic=False, n_jobs=-1, **kwargs):
         model_inference = algo(self)
 
         if stochastic:
-            data_unique_indexes = data.groupby(list(data.columns)).apply(
+            data_unique_indexes = data.groupby(list(data.columns), dropna=False).apply(
                 lambda t: t.index.tolist()
             )
             data_unique = data_unique_indexes.index.to_frame()
 
             pred_values = Parallel(n_jobs=n_jobs)(
                 delayed(model_inference.query)(
-                    variables=missing_variables,
-                    evidence=data_point.to_dict(),
+                    variables=missing_variables.union(
+                        set(data_point.index[data_point.isna()])
+                    ),
+                    evidence=data_point[~data_point.isna()].to_dict(),
                     show_progress=False,
                     **kwargs,
                 )
                 for index, data_point in tqdm(
                     data_unique.iterrows(), total=data_unique.shape[0]
                 )
             )
+
+            all_columns = data.columns.tolist() + [col for col in missing_variables]
             predictions = pd.DataFrame()
             for i, row in enumerate(data_unique_indexes):
-                p = pred_values[i].sample(n=len(row))
-                p.index = row
-                predictions = pd.concat((predictions, p), copy=False)
+                samples = pred_values[i].sample(n=len(row))
+                initial_variables = data_unique.iloc[[i]]
+                known_variables = initial_variables.dropna(axis=1)
+                known_df = known_variables.loc[known_variables.index.repeat(len(row))]
+                known_df.index = samples.index
+
+                complete_data = pd.concat([samples, known_df], axis="columns")
+                complete_data.index = row
+                complete_data = complete_data.reindex(columns=all_columns)
+                predictions = pd.concat((predictions, complete_data), copy=False)
 
-            return predictions.reindex(data.index)
+            return predictions.sort_index()
 
         else:
-            data_unique = data.drop_duplicates()
-            pred_values = []
+            data_unique_indexes = data.groupby(list(data.columns), dropna=False).apply(
```

Comment:
Because the log for both stochastic=True or False is now same, with the only difference that in one we fill the dataframe with samples and in the other with the value. Can we simplify the method here as there seems to be a lot of duplicated code?

Reply:
Understood, I'll reduce the unnecessary code.